### PR TITLE
Rework Crypto

### PR DESCRIPTION
**Pull request by Joannis**
- [x] Pure Swift for basic algorithms, as to not require OpenSSL for simple operations
- [x] PBKDF2
- [x] Basic hashes (streaming and normal interfaces)
- [x] Integration with Core/Streams
- [x] Base64 encoding
- [x] Base64 stream encoding
- [ ] Base64 decoding
- [ ] Base64 stream decoding
- [ ] Hexadecimal encoding stream
- [ ] Hexadecimal decoding stream

### CODE REVIEW

File: Sources/Crypto/Base64Decoder.swift [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+]
+
+fileprivate let encodeTable = [UInt8]("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".utf8)
+
+public final class Base64Encoder : Stream {
+    public typealias Input = ByteBuffer
+    public typealias Output = ByteBuffer
+    
+    public var outputStream: ((ByteBuffer) -> ())?
```
Comment (tanner0101) [inline Sources/Crypto/Base64Decoder.swift]: `public var outputStream: OutputHandler?` is more readable

File: Sources/Crypto/Base64Decoder.swift [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    public var errorStream: BaseStream.ErrorHandler?
```
Comment (tanner0101) [inline Sources/Crypto/Base64Decoder.swift]: no need for `BaseStream.` here

File: Sources/Crypto/Base64Decoder.swift [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    
+    public struct UnknownFailure : Error {}
```
Comment (tanner0101) [inline Sources/Crypto/Base64Decoder.swift]: we should have an actual error here. should conform to `Debuggable` and `Traceable` as well for nice debug messages. see: https://github.com/vapor/engine/blob/beta/Sources/TCP/Utilties/Error.swift for an example. there should probably be just one `Crypto.Error` used for everything.

File: Sources/Crypto/Base64Decoder.swift [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+    64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+]
+
+fileprivate let encodeTable = [UInt8]("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".utf8)
+
+public final class Base64Encoder : Stream {
```
Comment (tanner0101) [inline Sources/Crypto/Base64Decoder.swift]: comment

File: Sources/Crypto/Hash.swift [hunk 5]
```diff
@@ -0,0 +1,137 @@
+import Foundation
+
+public protocol Hash : class {
```
Comment (tanner0101) [inline Sources/Crypto/Hash.swift]: vapor style has been to have `:` touching the identifier.

```
public protocol Hash: class
```

File: Sources/Crypto/Hash.swift [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        }
+    }
+    
+    public func finalize(array: inout [UInt8]) {
+        return array.withUnsafeBufferPointer { buffer in
+            self.finalize(buffer)
+        }
+    }
+    
+    public func update(_ buffer: UnsafeBufferPointer<UInt8>) {
```
Comment (tanner0101) [inline Sources/Crypto/Hash.swift]: we should use `ByteBuffer` here

File: Sources/Crypto/HashHelpers.swift [hunk 7]
```diff
@@ -0,0 +1,5 @@
+extension Hash {
+    public static func hash(_ string: String) -> [UInt8] {
```
Comment (tanner0101) [inline Sources/Crypto/HashHelpers.swift]: I think we should use `Data` instead of `[UInt8]`

**Comment by Joannis** [comment]
Are the comments in crypto sufficient now? I can't really document the mathematical working on a hash. It wouldn't be useful for anyone maintaining either. I will place some links to the official specification of the algorithms, though

**Comment by tanner0101** [comment]
Closing in favor of #33

**Pull request closed without merging**